### PR TITLE
Add support for GNU BinaryConditionalOperator

### LINKED ISSUE #669: Add support for `?:` shorthand via `clang::OpaqueValueExpr` and `clang::BinaryConditionalOperator`
### Prerequisites

- [X] I have read the documentation and the proposed feature is not implemented.

### Description

Resolving this issue should result in the following not relying on the unsupported dialect: https://godbolt.org/z/YY7PKMEex

### PR DESCRIPTION
**Pull request by Jezurko**
Fix #669

### CODE REVIEW

File: include/vast/Dialect/HighLevel/HighLevelCF.td [hunk 1]
```diff
@@ -164,6 +164,39 @@ def HighLevel_CondOp : ControlFlowOp< "cond" >
...
+def HighLevel_BinaryCondOp
+  : HighLevel_Op<
+        "binary_cond",
+        [NoTerminator, DeclareOpInterfaceMethods<RegionKindInterface>]
+    >
+    , Results<(outs AnyType:$result)>
+{
+  let summary = "VAST binary conditional statement";
+  let description = [{
+    TODO
```
Comment (lkorenc) [inline include/vast/Dialect/HighLevel/HighLevelCF.td:176]: Fill.

File: lib/vast/CodeGen/DefaultStmtVisitor.cpp [hunk 2]
```diff
@@ -974,7 +992,29 @@ namespace vast::cg
             .freeze();
     }
 
-    // operation default_stmt_visitor::VisitOpaqueValueExpr(const clang::OpaqueValueExpr *expr)
+    operation default_stmt_visitor::VisitOpaqueValueExpr(const clang::OpaqueValueExpr *expr) {
+        const auto parents = actx.getParents< clang_stmt >(*expr);
```
Comment (xlauko) [inline lib/vast/CodeGen/DefaultStmtVisitor.cpp:996]: I think `expr` holds context you don't need to keep it in visitor.

File: lib/vast/CodeGen/DefaultStmtVisitor.cpp [hunk 3]
```diff
@@ -974,7 +992,29 @@ namespace vast::cg
...
+        // In C++ an AST stmt can have multiple parents
+        for (const auto &parent : parents) {
+
+            if (mlir::isa< clang::BinaryConditionalOperator >(parent.get< clang_stmt >())) {
+                //These arguments should be constructed when creating the BinaryCondOp
```
Comment (xlauko) [inline lib/vast/CodeGen/DefaultStmtVisitor.cpp:1001]:
```suggestion
                // These arguments should be constructed when creating the BinaryCondOp
```

File: lib/vast/Dialect/HighLevel/HighLevelOps.cpp [hunk 4]
```diff
@@ -521,35 +521,60 @@ namespace vast::hl
         build_region(bld, st, else_builder);
         st.addTypes(type);
     }
+    void BinaryCondOp::build(
+        Builder &bld, State &st, Type type,
+        builder_callback_ref opaq,
```
Comment (xlauko) [inline lib/vast/Dialect/HighLevel/HighLevelOps.cpp:526]:
```suggestion
        builder_callback_ref common,
```

File: lib/vast/Dialect/HighLevel/HighLevelOps.cpp [hunk 5]
```diff
@@ -521,35 +521,60 @@ namespace vast::hl
...
+        builder_callback_ref cond,
+        builder_callback_ref then_builder,
+        builder_callback_ref else_builder
+    ) {
+        InsertionGuard guard(bld);
+        build_region(bld, st, opaq);
```
Comment (xlauko) [inline lib/vast/Dialect/HighLevel/HighLevelOps.cpp:532]:
```suggestion
        build_region(bld, st, common);
```

File: lib/vast/Dialect/HighLevel/HighLevelOps.cpp [hunk 6]
```diff
@@ -521,35 +521,60 @@ namespace vast::hl
         build_region(bld, st, else_builder);
         st.addTypes(type);
     }
+    void BinaryCondOp::build(
+        Builder &bld, State &st, Type type,
+        builder_callback_ref opaq,
+        builder_callback_ref cond,
```
Comment (xlauko) [inline lib/vast/Dialect/HighLevel/HighLevelOps.cpp:527]:
```suggestion
        builder_callback_ref cond_builder,
```

File: lib/vast/Dialect/HighLevel/HighLevelOps.cpp [hunk 7]
```diff
@@ -521,35 +521,60 @@ namespace vast::hl
...
+        builder_callback_ref then_builder,
+        builder_callback_ref else_builder
+    ) {
+        InsertionGuard guard(bld);
+        build_region(bld, st, opaq);
+        build_region(bld, st, cond);
```
Comment (xlauko) [inline lib/vast/Dialect/HighLevel/HighLevelOps.cpp:533]:
```suggestion
        build_region(bld, st, cond_builder);
```

**Review by xlauko — APPROVED** [review]

**Pull request merged**